File: threadly-spring-boot/src/main/java/com/threadly/comment/domain/Comment.java
package com.threadly.comment.domain;

import com.threadly.comment.CreateCommentRequest;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import java.time.Instant;
import java.util.Map;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

@Entity
@Table(name = "comments")
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Getter
@Setter
public class Comment {

  @Id
  @GeneratedValue(strategy = GenerationType.UUID)
  private UUID id;

  @Column
  private UUID userId;

  @Column
  private UUID postId;

  @Column
  private UUID communityId;

  @Column
  private Integer depth;

  @Column
  private UUID parentId;

  @Column
  private Integer childCount;

  @Column(columnDefinition = "jsonb", nullable = false)
  @JdbcTypeCode(SqlTypes.JSON)
  private Map<String, Object> contentJson;

  @Column(columnDefinition = "text")
  private String contentHtml;

  @Column(columnDefinition = "text")
  private String contentText;

  @CreationTimestamp
  @Column(updatable = false, nullable = false)
  private Instant createdAt;

  @UpdateTimestamp
  @Column(nullable = false)
  private Instant updatedAt;

  @Version
  @Column(nullable = false)
  private Long version;

  public static Comment from(CreateCommentRequest createCommentRequest) {
    return Comment.builder()
        .depth(createCommentRequest.depth())
        .postId(createCommentRequest.postId())
        .parentId(createCommentRequest.parentId())
        .communityId(createCommentRequest.communityId())
        .contentJson(createCommentRequest.contentJson())
        .contentHtml(createCommentRequest.contentHtml())
        .contentText(createCommentRequest.contentText())
        .userId(createCommentRequest.actorId())
          .childCount(0)
        .build();
  }

}

File: threadly-ui/src/types/comment.ts
import z from "zod";
import { instant, sliceSchema } from "./utils";

export const createCommentRequestSchema = z.object({
	depth: z.number(),
	postId: z.string(),
	actorId: z.string(),
	communityId: z.string(),
	parentId: z.string().nullable().default(null),
	contentJson: z.any(),
	contentHtml: z.string(),
	contentText: z.string(),
});
export const commentDTOSchema = z.object({
	id: z.string(),
	userId: z.string(),
	postId: z.string(),
	communityId: z.string(),
	depth: z.number(),
	parentId: z.string().optional(),
	contentJson: z.any(),
	contentHtml: z.string(),
	contentText: z.string(),
	childCount: z.number(),
	createdAt: instant,
	updatedAt: instant,
});
export const commentDTOSliceSchema = sliceSchema(commentDTOSchema);
export type CommentDTO = z.infer<typeof commentDTOSchema>;
export type CommentDTOSlice = z.infer<typeof commentDTOSliceSchema>;
export type CreateCommentRequest = z.infer<typeof createCommentRequestSchema>;
